### ISSUE TITLE
C API Support

### ISSUE THREAD

**Issue by tenderlove** [emcrisostomo/fswatch#120]
Hi,

I'm having troubles with the C API.  It seems like the C targeted header files are still trying to pull in C++ headers.  For example [`<ctime>` isn't available](https://github.com/emcrisostomo/fswatch/blob/master/libfswatch/src/libfswatch/c/cmonitor.h#L29), and neither is [`<climits>`](https://github.com/emcrisostomo/fswatch/blob/a79113ac6635620bd8e70565b5f6acfec8dc4e00/libfswatch/src/libfswatch/c/cevent.h#L30-L31).  There are also some various compiler warnings (like missing enums in declarations, missing `bool` types and that [C doesn't support default params](https://github.com/emcrisostomo/fswatch/blob/a79113ac6635620bd8e70565b5f6acfec8dc4e00/libfswatch/src/libfswatch/c/libfswatch.h#L101)).

I'm using OS X, and working with the version installed by Homebrew (1.9.2).  I would send a patch, but I'm also having a hard time getting the library to compile from git.  I've verified I have the correct versions of GNU autotools / libtool installed, but I get an error when running configure:

```
[aaron@TC fswatch (master)]$ ./configure
configure: error: cannot find install-sh, install.sh, or shtool in config "."/config
[aaron@TC fswatch (master)]$
```

Any help would be appreciated.

**Comment by emcrisostomo** [comment]
Hi @tenderlove,

I don't know about the `configure` error: it sounds like something is missing.  I'm using OS X as main development platform for `fswatch` and that shouldn't happen.  Did you `autogen.sh` the sources or started from a [source distribution][source]?

As far as the other errors and warnings, I can fix them, but I'd like to see exactly the errors you're getting. It's true that the it's using C++ names of some C headers, and that should be fixed.  For starters, I'll have a look at the two you've sent (`ctime` and `climits`).

[source]: https://github.com/emcrisostomo/fswatch/releases/download/1.9.2/fswatch-1.9.2.tar.gz

**Comment by emcrisostomo** [comment]
@tenderlove, would you please check out the `hotfix/1.9.3` branch and try it?

**Comment by smukkecharlie** [comment]
Hi there, I just tested the 1.9.3 branch against clang gcc on OSx. There still are some minor compilation issues. Listed below. For full details look [here](https://dotdideriksen.blogspot.se/2016/06/git-branching-fswatch-to-test-c.html)

$ gcc -I /usr/local/include -o "fswatch_test" fswatch_test.c /usr/local/lib/libfswatch.a 
In file included from fswatch_test.c:3:
In file included from /usr/local/include/libfswatch/c/libfswatch.h:33:
/usr/local/include/libfswatch/c/cevent.h:94:59: error: must use 'enum' tag to refer to type
      'fsw_event_flag'
  FSW_STATUS fsw_get_event_flag_by_name(const char *name, fsw_event_flag *flag);
                                                          ^
/usr/local/include/libfswatch/c/cevent.h:105:39: error: must use 'enum' tag to refer to type
      'fsw_event_flag'
  char *fsw_get_event_flag_name(const fsw_event_flag flag);
                                      ^
/usr/local/include/libfswatch/c/cevent.h:118:5: error: must use 'enum' tag to refer to type
      'fsw_event_flag'
    fsw_event_flag * flags;
    ^
In file included from fswatch_test.c:3:
In file included from /usr/local/include/libfswatch/c/libfswatch.h:35:
/usr/local/include/libfswatch/c/cfilter.h:47:5: error: must use 'enum' tag to refer to type
      'fsw_filter_type'
    fsw_filter_type type;
    ^
/usr/local/include/libfswatch/c/cfilter.h:57:5: error: must use 'enum' tag to refer to type
      'fsw_event_flag'
    fsw_event_flag flag;
    ^
In file included from fswatch_test.c:3:
/usr/local/include/libfswatch/c/libfswatch.h:102:37: error: must use 'enum' tag to refer to
      type 'fsw_monitor_type'
  FSW_HANDLE fsw_init_session(const fsw_monitor_type type = system_default_monitor_type);
                                    ^
/usr/local/include/libfswatch/c/libfswatch.h:102:59: error: C does not support default
      arguments
  FSW_HANDLE fsw_init_session(const fsw_monitor_type type = system_default_monitor_type);
                                                          ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~
7 errors generated.

**Comment by johndid** [comment]
Once the compilation errors are fixed, the linker (ld) reports many issues with missing c++ objects. Would you like me to report these issues here? or commit my changes to the branch?

**Comment by emcrisostomo** [comment]
Hi @johndid, thanks for the work you're doing: the C library really needed this.  I will review all the information you already provided.  If you prefer contributing your changes, feel free to fork `fswatch`: since there are already many changes, it would be better to open a `feature` branch (have a look at [this][flow] if you're new to `git flow`).

[flow]: http://nvie.com/posts/a-successful-git-branching-model/

**Comment by johndid** [comment]
You're welcome @emcrisostomo. I'll have a look at the process stuff you linked above and see what I can do about it.

**Comment by muyinliu** [comment]
Thanks, @johndid. Your demo [fswatch_test.c](https://github.com/johndid/playground/blob/master/c/fswatch_test.c) is really helpful.

I'm writing a CFFI(Common Lisp) wrap cl-fswatch to libfswatch, few days later it will be upload to GitHub.

**Comment by emcrisostomo** [comment]
Hi @johndid, thanks for the pull request: I've just merged it.  I will test it with the `fswatch_test.c` file you contributed and if it's ok I will publish `1.9.3`.

**Comment by emcrisostomo** [comment]
Fixed in [1.9.3][release].

[release]: https://github.com/emcrisostomo/fswatch/releases/tag/1.9.3

**Comment by johndid** [comment]
Glad I could help ;)
